### PR TITLE
NFTs - updating metadata and new puzzles 

### PR DESCRIPTION
**Pull request by trepca**

### CODE REVIEW

File: chia/rpc/wallet_rpc_api.py [hunk 1]
```diff
@@ -1339,22 +1340,33 @@ async def nft_get_nfts(self, request) -> Dict:
 
     async def nft_transfer_nft(self, request):
         assert self.service.wallet_state_manager is not None
-        wallet_id = int(request["wallet_id"])
+        wallet_id = uint32(request["wallet_id"])
```
Comment (paninaro) [inline chia/rpc/wallet_rpc_api.py:1343]: We can track this separately, but we still need to add support for specifying a fee to push the transaction through.

File: chia/wallet/puzzles/nft_metadata_updater_default.clvm.hex [hunk 2]
```diff
@@ -0,0 +1 @@
+ff02ffff01ff04ffff04ffff02ffff03ff17ffff01ff02ff02ffff04ff02ffff04ff05ffff04ff17ff8080808080ffff010580ff0180ffff04ff0bff808080ffff01ff808080ffff04ffff01ff02ffff03ff05ffff01ff02ffff03ffff09ff11ffff017580ffff01ff04ffff04ffff0175ffff04ff0bff198080ff0d80ffff01ff04ff09ffff02ff02ffff04ff02ffff04ff0dffff04ff0bff80808080808080ff0180ff8080ff0180ff018080
```
Comment (ytx1991) [inline chia/wallet/puzzles/nft_metadata_updater_default.clvm.hex:1]: You may need to add this field in the clvm compilation tests

File: chia/wallet/nft_wallet/nft_wallet.py [hunk 3]
```diff
@@ -530,9 +528,41 @@ async def transfer_nft(
             name=bytes32(token_bytes()),
             memos=list(compute_memos(full_spend).items()),
         )
-        await self.standard_wallet.push_transaction(nft_record)
-        full_spend.debug()
-        return full_spend
+        return nft_record
+
+    async def update_metadata(self, nft_coin_info: NFTCoinInfo, uri: str) -> Optional[SpendBundle]:
+        coin = nft_coin_info.coin
```
Comment (ytx1991) [inline chia/wallet/nft_wallet/nft_wallet.py:534]: Optional comment, may be distinguish  this with NFT1 approach. Like add a if statement or change the method name to update_metadata_nft0

File: chia/wallet/puzzles/nft_state_layer.clvm [hunk 4]
```diff
@@ -27,21 +27,25 @@
...
+        )
+        (if (> (f (f conditions)) 0)
+          (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (list (list METADATA METADATA_UPDATER_PUZZLE_HASH) (r conditions)) my_amount new_puzhash))
+          (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (list (list METADATA METADATA_UPDATER_PUZZLE_HASH) (r conditions)) my_amount new_puzhash)
         )
-        (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH METADATA_UPDATER_PUZZLE_HASH (list METADATA (r conditions)) my_amount))
       )
-      ()
+      (if new_puzhash
+        (list (list CREATE_COIN (nft_state_layer_puzzle_hash NFT_STATE_LAYER_MOD_HASH METADATA METADATA_UPDATER_PUZZLE_HASH (f (r new_puzhash)))
```
Comment (ytx1991) [inline chia/wallet/puzzles/nft_state_layer.clvm:43]: Align the following lines

File: chia/wallet/nft_wallet/nft_wallet.py [hunk 5]
```diff
@@ -165,6 +164,12 @@ async def get_max_send_amount(self, records=None):
         """This is the confirmed balance, which we set to 0 as the NFT wallet doesn't have one."""
         return uint128(0)
 
+    def get_nft_coin_by_id(self, nft_coin_id) -> NFTCoinInfo:
```
[the remark this answers is not in the document]
Comment (trepca) [reply chia/wallet/nft_wallet/nft_wallet.py:167]: nice catch

File: chia/wallet/nft_wallet/nft_wallet.py [hunk 6]
```diff
@@ -457,12 +462,15 @@ async def sign(self, spend_bundle: SpendBundle) -> SpendBundle:
...
-                )
-
+                try:
+                    error, conditions, cost = conditions_dict_for_solution(
+                        spend.puzzle_reveal.to_program(),
+                        spend.solution.to_program(),
+                        self.wallet_state_manager.constants.MAX_BLOCK_COST_CLVM,
+                    )
+                except ValueError:
+                    # ignore invalid condition for updating metadata
```
[the remark this answers is not in the document]
Comment (trepca) [reply chia/wallet/nft_wallet/nft_wallet.py:472]: actually, this can't happen anymore as we remove magic conditions in chialisp, will remove

File: chia/wallet/puzzles/nft_state_layer.clvm [hunk 7]
```diff
@@ -55,16 +65,14 @@
...
         (if (= (sha256tree1 (f (r (f conditions_loop)))) METADATA_UPDATER_PUZZLE_HASH)
-          ; apply the metadata updater
-          (a (f (r (f conditions_loop))) (list CURRENT_METADATA (f (r (r (f conditions_loop))))))
+          (a (f (r (f conditions_loop))) (list CURRENT_METADATA METADATA_UPDATER_PUZZLE_HASH (f (r (r (f conditions_loop))))))
           (x)
         )
         (check_for_metadata_prog_reveal METADATA_UPDATER_PUZZLE_HASH CURRENT_METADATA (r conditions_loop))
       )
-      (list CURRENT_METADATA 0)
+      (list (list CURRENT_METADATA METADATA_UPDATER_PUZZLE_HASH) 0)
```
Comment (matt-o-how) [inline chia/wallet/puzzles/nft_state_layer.clvm:75]: This is the design pattern I'm least sure about

File: chia/wallet/nft_wallet/nft_wallet.py (hunk 3 above)
Comment (trepca) [reply chia/wallet/nft_wallet/nft_wallet.py:530]: ok, makes sense, will add support this in a separate PR

**Review by paninaro — APPROVED** [review]
I'm not familiar with the chialisp changes, but the rest looks good to me.

File: chia/wallet/puzzles/nft_state_layer.clvm [hunk 8]
```diff
@@ -27,21 +27,31 @@
...
-            (c CREATE_COIN
-               (c (nft_state_layer_puzzle_hash NFT_STATE_LAYER_MOD_HASH METADATA METADATA_UPDATER_PUZZLE_HASH (f (r (f conditions))))
-                  (c my_amount
-                     (r (r (r (f conditions)))))))
-              (r conditions))
-          (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH METADATA_UPDATER_PUZZLE_HASH (list METADATA (r conditions)) my_amount))
+          (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (list (list METADATA METADATA_UPDATER_PUZZLE_HASH) (r conditions)) my_amount (f conditions))
+          (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (list (list METADATA METADATA_UPDATER_PUZZLE_HASH) (r conditions)) my_amount new_create_coin_condition))
+        )
+        (if (> (f (f conditions)) 0)
```
[the remark this answers is not in the document]
Comment (ytx1991) [reply chia/wallet/puzzles/nft_state_layer.clvm:37]: I'm trying to understand but TBH, hard to follow the idea.
